src/QueryRenderer.ts

```diff
@@ -239,27 +239,26 @@ class QueryRenderChild extends MarkdownRenderChild {
             const footnotes = listItem.querySelectorAll('[data-footnote-id]');
             footnotes.forEach((footnote) => footnote.remove());
 
-            const shortMode = this.query.layoutOptions.shortMode;
-
             const extrasSpan = listItem.createSpan('task-extras');
 
-            if (!this.query.layoutOptions.hidePostponeButton) {
-                this.addPostponeButton(extrasSpan, task, shortMode);
-            }
-
             if (!this.query.layoutOptions.hideUrgency) {
                 this.addUrgency(extrasSpan, task);
             }
 
+            const shortMode = this.query.layoutOptions.shortMode;
+
             if (!this.query.layoutOptions.hideBacklinks) {
-                const shortMode = this.query.layoutOptions.shortMode;
                 this.addBacklinks(extrasSpan, task, shortMode, isFilenameUnique);
             }
 
             if (!this.query.layoutOptions.hideEditButton) {
                 this.addEditButton(extrasSpan, task);
             }
 
+            if (!this.query.layoutOptions.hidePostponeButton && !task.isDone) {
```

Comment:
### Don't depend on human testing - and testers knowing what to test

This change is adding logic - `&& !task.isDone` - but no automated tests for the logic.

As a result, it's requiring current and future developers to:

1. build and test the behaviour whenever changing this code
2. actually know what the expected behaviour is, when manually testing.

Also, we have already planned several more changes to that logic, so the testing is going to get more complicated, and each time we add new logic, we are going to need to manually test all the different combinations of behaviours.

Better to say this kind of thing:

```js
    if (!this.query.layoutOptions.hidePostponeButton && shouldShowPostponeButton(task) {
```

And have `shouldShowPostponeButton()` - or whatever you decide to call it - be in a non-gui file, and have tests.

And the test would demonstrate the behaviour for a task of each of the 5 status types.

---

### Location of extracted code

I know that we extracted the initial postpone-related helper functions in to `QueryRendererHelper.test.ts`. I've been reflecting on that, it might as well be called 'bucket of random functions'... and it really falls foul of:

- move related stuff closer together
- and unrelated stuff further away

I suggest creating `Postponer.ts` and corresponding test file - and putting it alongside `TasksDate` for now.

Reply:
Many thanks for the detailed comment.

> And have `shouldShowPostponeButton()` - or whatever you decide to call it - be in a non-gui file, and have tests.

Done.

> And the test would demonstrate the behaviour for a task of each of the 5 status types.

I have troubles writing the tests since I lack knowledge of Status framework so I'd better not write them for now rather than do something weird.

> I suggest creating `Postponer.ts` and corresponding test file - and putting it alongside `TasksDate` for now.

Nice, thanks. Done.

Reply:
> I have troubles writing the tests since I lack knowledge of Status framework so I'd better not write them for now rather than do something weird.

I starting typing up how to do it, and it did get a bit long, so sure - let's talk it through on Tuesday.

Reply:
Maybe add a TODO in the tests, to remind us to come back and add a test for this function...

Reply:
Please consider writing tests and pushing to my branch if that is quicker than typing explanations here.

Reply:
OK, working on that now.

Reply:
Done in 2a826fc90dcc0d0cf8f4e8af5e74c2e4a8f2517a

Reply:
Thanks

Reply:
I was missing this part actually:

```ts
const status = new Status(new StatusConfiguration('p', 'Test', 'q', true, statusType));
```